Spatial.gd

```diff
@@ -757,6 +749,97 @@ func clear_adjustment_nodes():
+	refresh_path2D_points(path_index)
+
+func reverse_trail(path_index: int):
+	var trail_data = self.currently_active_category.get_metadata(0).category3d.paths[path_index].waypoint.get_trail_data()
+	trail_data.get_points_x().invert()
+	trail_data.get_points_y().invert()
+	trail_data.get_points_z().invert()
+	refresh_path3D_points(path_index)
+	refresh_path2D_points(path_index)
+
+func get_trail_point_count(path_index: int):
+	var trail_data = self.currently_active_category.get_metadata(0).category3d.paths[path_index].waypoint.get_trail_data()
+	return len(trail_data.get_points_x())
+
+func get_trail_point_position(path_index: int, point_index: int):
+	var position: Vector3
+	var trail_data = self.currently_active_category.get_metadata(0).category3d.paths[path_index].waypoint.get_trail_data()
+	position[0] = trail_data.get_points_x()[point_index]
+	position[1] = trail_data.get_points_y()[point_index]
+	position[2] = trail_data.get_points_z()[point_index]
+	return position
+
+func add_trail_point(position: Vector3, path_index: int, point_index: int = -1):
+	var trail_data = self.currently_active_category.get_metadata(0).category3d.paths[path_index].waypoint.get_trail_data()
+	if point_index == -1:
+		trail_data.get_points_x().append(position.x)
+		trail_data.get_points_y().append(position.y)
+		trail_data.get_points_z().append(-position.z)
+	else:
+		trail_data.get_points_x().insert(point_index, position.x)
+		trail_data.get_points_y().insert(point_index, position.y)
+		trail_data.get_points_z().insert(point_index, -position.z)
+	refresh_path3D_points(path_index)
+	refresh_path2D_points(path_index)
+
+func remove_trail_point(position: Vector3, path_index: int, point_index: int = -1):
+	var trail_data = self.currently_active_category.get_metadata(0).category3d.paths[path_index].waypoint.get_trail_data()
+	trail_data.get_points_x().remove(position.x)
+	trail_data.get_points_y().remove(position.y)
+	trail_data.get_points_z().remove(-position.z)
+	refresh_path3D_points(path_index)
+	refresh_path2D_points(path_index)
+
+func new_trail_point_after(path_index: int, point_index: int):
+	var start: Vector3 = get_trail_point_position(path_index, point_index)
+	var target_position: Vector3
+	if get_trail_point_count(path_index) > point_index+1:
+		var end: Vector3 = get_trail_point_position(path_index, point_index+1)
+		target_position = ((start-end)/2) + end
+	else:
+		target_position = Vector3(self.player_position.x, self.player_position.y, -self.player_position.z)
+	add_trail_point(target_position, point_index+1)
+
+func refresh_path3D_points(path_index: int):
+	var path_points := PoolVector3Array()
+	var path3d = self.currently_active_category.get_metadata(0).category3d.paths[path_index]
+	var trail_data = path3d.waypoint.get_trail_data()
+	for index in range(0, trail_data.get_points_z().size()):
+		path_points.append(Vector3(trail_data.get_points_x()[index], trail_data.get_points_y()[index], -trail_data.get_points_z()[index]))
```

Comment:
Are you rebuilding a proto array into a poolvector3 array to pass into path3d's refresh_mesh which rebuilds a poolvector3 array into the mesh? That feels like it can be streamlined.

Reply:
Good point. Changed to bypass the poolvector3 array